server.py

```diff
@@ -1598,19 +1598,40 @@ async def wall_card(ctx: ContextType):
 
     await ctx.reply(f"Current card in the {config.baalorbot.prefix}hole in the wall for the ladder savefile: {p.hole_card}")
 
+@command("range", flag="me")
+async def set_run_stats_by_date(ctx: ContextType, date_string: str):
+    """Update the default range for the run stats by date, this is separate from the all-time stats run cache."""
+    try:
+        date_tuple = parse_date_range(date_string)
+    except:
+        await ctx.reply("Invalid date string. Use YYYY-MM-DD-YYYY-MM-DD (MM and DD optional), YYYY-MM-DD+ (no end date), YYYY-MM-DD- (no start date)")
+        return
+    update_range(date_tuple[0], date_tuple[1])
+    await ctx.reply("Run stats have been updated for the given range")
+
 @command("kills", "wins")
-async def calculate_wins_cmd(ctx: ContextType):
-    """Display the cumulative number of wins for the year-long challenge."""
-    run_stats = get_run_stats()
-    msg = "A20 Heart kills in {0.current_year}: Total: {1.all_character_count} - Ironclad: {1.ironclad_count} - Silent: {1.silent_count} - Defect: {1.defect_count} - Watcher: {1.watcher_count}"
-    await ctx.reply(msg.format(run_stats, run_stats.year_wins[run_stats.current_year]))
+async def calculate_wins_cmd(ctx: ContextType, *date_string: str):
+    """Display the cumulative number of wins for an optional date range."""
+    msg = "A20 Heart kills ({0.date_range_string}): Total: {1.all_character_count} - Ironclad: {1.ironclad_count} - Silent: {1.silent_count} - Defect: {1.defect_count} - Watcher: {1.watcher_count}"
+    await _send_standard_run_stats_message(ctx, msg, "all_wins", date_string)
 
 @command("losses")
-async def calculate_losses_cmd(ctx: ContextType):
-    """Display the cumulative number of losses for the year-long challenge."""
-    msg = "A20 Heart losses in {0.current_year}: Total: {1.all_character_count} - Ironclad: {1.ironclad_count} - Silent: {1.silent_count} - Defect: {1.defect_count} - Watcher: {1.watcher_count}"
-    run_stats = get_run_stats()
-    await ctx.reply(msg.format(run_stats, run_stats.year_losses[run_stats.current_year]))
+async def calculate_losses_cmd(ctx: ContextType, *date_string: str):
+    """Display the cumulative number of losses for an optional date range."""
+    msg = "A20 Heart losses ({0.date_range_string}): Total: {1.all_character_count} - Ironclad: {1.ironclad_count} - Silent: {1.silent_count} - Defect: {1.defect_count} - Watcher: {1.watcher_count}"
+    await _send_standard_run_stats_message(ctx, msg, "all_losses", date_string)
+
+async def _send_standard_run_stats_message(ctx: ContextType, msg: str, prop_name: str, date_string: tuple[str, ...]):
+    run_stats = None
+    if len("".join(date_string)) == 0:
```

Comment:
let me know if this len check is a decent way to see if the user sent any dates, since it's optional. I suppose I could just straight do len(date_string) instead of joining?

Reply:
I would actually recommend making the type `Optional[str]`, with the command args having a default of `None`. Then you can check for None instead. Checking for a length of 0 can almost always be improved somewhere!